File: python/src/mcp_test_kits/auth/pkce.py
"""PKCE (Proof Key for Code Exchange) verification."""

from __future__ import annotations

import base64
import hashlib


def verify_code_challenge(verifier: str, challenge: str) -> bool:
    """Verify PKCE S256 code challenge.

    Args:
        verifier: Code verifier provided during token exchange.
        challenge: Code challenge from authorization request.

    Returns:
        True if challenge matches verifier, False otherwise.
    """
    # Compute SHA256 hash of verifier
    digest = hashlib.sha256(verifier.encode("ascii")).digest()

    # Base64 URL-encode without padding
    computed_challenge = base64.urlsafe_b64encode(digest).rstrip(b"=").decode("ascii")

    return computed_challenge == challenge

File: typescript/src/auth/pkce.ts
/**
 * PKCE (Proof Key for Code Exchange) verification
 */

import { createHash } from "crypto";

/**
 * Verify PKCE S256 code challenge
 *
 * @param verifier - Code verifier provided during token exchange
 * @param challenge - Code challenge from authorization request
 * @returns True if challenge matches verifier, false otherwise
 */
export function verifyCodeChallenge(
  verifier: string,
  challenge: string,
): boolean {
  // Compute SHA256 hash of verifier
  const hash = createHash("sha256").update(verifier).digest();

  // Base64 URL-encode without padding
  const computedChallenge = hash
    .toString("base64")
    .replace(/\+/g, "-")
    .replace(/\//g, "_")
    .replace(/=/g, "");

  return computedChallenge === challenge;
}
